api/indexer/indexer.go

```diff
@@ -8,230 +8,252 @@ import (
+	iter := i.blockDB.NewIteratorWithPrefix(blockEntryKeyPrefix)
 	defer iter.Release()
 
-	var lastHeight uint64
 	for iter.Next() {
 		value := iter.Value()
 		blk, err := chain.UnmarshalExecutedBlock(value, i.parser)
 		if err != nil {
 			return err
 		}
-
-		i.blockIDToHeight.Put(blk.Block.GetID(), blk.Block.Hght)
-		i.blockHeightToBlock.Put(blk.Block.Hght, blk)
-		lastHeight = blk.Block.Hght
+		i.insertBlockIntoCache(blk)
 	}
 	if err := iter.Error(); err != nil {
 		return err
 	}
 	iter.Release()
-
-	i.lastHeight.Store(lastHeight)
-
-	if lastHeight > i.blockWindow {
-		lastRetainedHeight := lastHeight - i.blockWindow
-		lastRetainedHeightBytes := binary.BigEndian.AppendUint64(nil, lastRetainedHeight)
-		if err := i.blockDB.DeleteRange(nil, lastRetainedHeightBytes); err != nil {
-			return err
-		}
-	}
-
 	return nil
 }
 
 func (i *Indexer) Notify(_ context.Context, blk *chain.ExecutedBlock) error {
-	if err := i.storeTransactions(blk); err != nil {
-		return err
+	if i.blockWindow == 0 {
+		return nil
 	}
+
+	i.mu.Lock()
+	i.insertBlockIntoCache(blk)
+	i.mu.Unlock()
+
 	return i.storeBlock(blk)
 }
 
-func (i *Indexer) storeBlock(blk *chain.ExecutedBlock) error {
-	if i.blockWindow == 0 {
-		return nil
+// insertBlockIntoCache add the given block and it's transaction
+// to the indexer's cache.
+// assume : the write lock was already obtained by the caller.
+func (i *Indexer) insertBlockIntoCache(blk *chain.ExecutedBlock) {
+	i.cachedBlocks.Push(blk.Block.Hght)
+	i.blockIDToHeight[blk.Block.GetID()] = blk.Block.Hght
+	i.blockHeightToBlock[blk.Block.Hght] = blk
+
+	for idx, tx := range blk.Block.Txs {
+		i.txCache[tx.GetID()] = cachedTransaction{
+			blk:   blk,
+			index: idx,
+		}
 	}
+	i.lastHeight = blk.Block.Hght
+}
 
+// storeBlock persist the given block to the database, and remove the
+// evicted blocks.
+func (i *Indexer) storeBlock(blk *chain.ExecutedBlock) error {
 	executedBlkBytes, err := blk.Marshal()
 	if err != nil {
 		return err
 	}
 
-	if err := i.blockDB.Put(binary.BigEndian.AppendUint64(nil, blk.Block.Hght), executedBlkBytes); err != nil {
+	blkBatch := i.blockDB.NewBatch()
+	defer blkBatch.Reset()
+
+	if err := blkBatch.Put(blockEntryKey(blk.Block.Hght), executedBlkBytes); err != nil {
 		return err
 	}
-	// Ignore overflows in key calculation which will simply delete a non-existent key
-	if err := i.blockDB.Delete(binary.BigEndian.AppendUint64(nil, blk.Block.Hght-i.blockWindow)); err != nil {
+
+	for _, evictedBlockHeight := range i.evictedBlocks {
+		if err := blkBatch.Delete(blockEntryKey(evictedBlockHeight)); err != nil {
+			return err
+		}
+	}
+
+	if err := blkBatch.Put(latestBlockHeightKey, binary.BigEndian.AppendUint64(nil, i.lastHeight)); err != nil {
 		return err
 	}
 
-	i.blockIDToHeight.Put(blk.Block.GetID(), blk.Block.Hght)
-	i.blockHeightToBlock.Put(blk.Block.Hght, blk)
-	i.lastHeight.Store(blk.Block.Hght)
+	// clear the slice without allocation
+	i.evictedBlocks = i.evictedBlocks[:0]
```

Comment:
Do we need to grab the lock here?

I think the answer is just barely no because this is only accessed by the writer and not on reads and the writer is single threaded, but this is fairly non-obvious.